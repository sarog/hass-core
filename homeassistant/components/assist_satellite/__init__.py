"""Base class for assist satellite entities."""

import logging

import voluptuous as vol

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers import config_validation as cv
from homeassistant.helpers.entity_component import EntityComponent
from homeassistant.helpers.typing import ConfigType

from .const import DOMAIN, AssistSatelliteEntityFeature
from .entity import AssistSatelliteEntity, AssistSatelliteEntityDescription
from .errors import SatelliteBusyError
from .websocket_api import async_register_websocket_api

__all__ = [
    "DOMAIN",
    "AssistSatelliteEntity",
    "AssistSatelliteEntityDescription",
    "AssistSatelliteEntityFeature",
    "SatelliteBusyError",
]

_LOGGER = logging.getLogger(__name__)

PLATFORM_SCHEMA_BASE = cv.PLATFORM_SCHEMA_BASE


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    component = hass.data[DOMAIN] = EntityComponent[AssistSatelliteEntity](
        _LOGGER, DOMAIN, hass
    )
    await component.async_setup(config)

    component.async_register_entity_service(
        "announce",
        vol.All(
            cv.make_entity_service_schema(
                {
                    vol.Optional("message"): str,
                    vol.Optional("media_id"): str,
                }
            ),
            cv.has_at_least_one_key("message", "media_id"),
        ),
        "async_internal_announce",
        [AssistSatelliteEntityFeature.ANNOUNCE],
    )
    async_register_websocket_api(hass)

    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up a config entry."""
    component: EntityComponent[AssistSatelliteEntity] = hass.data[DOMAIN]
    return await component.async_setup_entry(entry)


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    component: EntityComponent[AssistSatelliteEntity] = hass.data[DOMAIN]
    return await component.async_unload_entry(entry)
