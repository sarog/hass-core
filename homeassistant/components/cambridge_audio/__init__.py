"""The Cambridge Audio integration."""

from __future__ import annotations

import asyncio

from aiostreammagic import StreamMagicClient

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_HOST, Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady

from .const import CONNECT_TIMEOUT, STREAM_MAGIC_EXCEPTIONS

PLATFORMS: list[Platform] = [Platform.MEDIA_PLAYER]

type CambridgeAudioConfigEntry = ConfigEntry[StreamMagicClient]


async def async_setup_entry(
    hass: HomeAssistant, entry: CambridgeAudioConfigEntry
) -> bool:
    """Set up Cambridge Audio integration from a config entry."""

    client = StreamMagicClient(entry.data[CONF_HOST])

    try:
        async with asyncio.timeout(CONNECT_TIMEOUT):
            await client.connect()
    except STREAM_MAGIC_EXCEPTIONS as err:
        raise ConfigEntryNotReady(f"Error while connecting to {client.host}") from err
    entry.runtime_data = client

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(
    hass: HomeAssistant, entry: CambridgeAudioConfigEntry
) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        await entry.runtime_data.disconnect()
    return unload_ok
