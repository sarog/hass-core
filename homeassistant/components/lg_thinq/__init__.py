"""Support for LG ThinQ Connect device."""

from __future__ import annotations

import asyncio
import logging

from thinqconnect import ThinQApi, ThinQAPIException
from thinqconnect.integration import async_get_ha_bridge_list

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_ACCESS_TOKEN, CONF_COUNTRY, Platform
from homeassistant.core import HomeAssistant, callback
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import CONF_CONNECT_CLIENT_ID
from .coordinator import DeviceDataUpdateCoordinator, async_setup_device_coordinator

type ThinqConfigEntry = ConfigEntry[dict[str, DeviceDataUpdateCoordinator]]

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SWITCH]

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
    """Set up an entry."""
    entry.runtime_data = {}

    access_token = entry.data[CONF_ACCESS_TOKEN]
    client_id = entry.data[CONF_CONNECT_CLIENT_ID]
    country_code = entry.data[CONF_COUNTRY]

    thinq_api = ThinQApi(
        session=async_get_clientsession(hass),
        access_token=access_token,
        country_code=country_code,
        client_id=client_id,
    )

    # Setup coordinators and register devices.
    await async_setup_coordinators(hass, entry, thinq_api)

    # Set up all platforms for this device/entry.
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    # Clean up devices they are no longer in use.
    async_cleanup_device_registry(hass, entry)

    return True


async def async_setup_coordinators(
    hass: HomeAssistant,
    entry: ThinqConfigEntry,
    thinq_api: ThinQApi,
) -> None:
    """Set up coordinators and register devices."""
    # Get a list of ha bridge.
    try:
        bridge_list = await async_get_ha_bridge_list(thinq_api)
    except ThinQAPIException as exc:
        raise ConfigEntryNotReady(exc.message) from exc

    if not bridge_list:
        return

    # Setup coordinator per device.
    task_list = [
        hass.async_create_task(async_setup_device_coordinator(hass, bridge))
        for bridge in bridge_list
    ]
    task_result = await asyncio.gather(*task_list)
    for coordinator in task_result:
        entry.runtime_data[coordinator.unique_id] = coordinator


@callback
def async_cleanup_device_registry(hass: HomeAssistant, entry: ThinqConfigEntry) -> None:
    """Clean up device registry."""
    new_device_unique_ids = [
        coordinator.unique_id for coordinator in entry.runtime_data.values()
    ]
    device_registry = dr.async_get(hass)
    existing_entries = dr.async_entries_for_config_entry(
        device_registry, entry.entry_id
    )

    # Remove devices that are no longer exist.
    for old_entry in existing_entries:
        old_unique_id = next(iter(old_entry.identifiers))[1]
        if old_unique_id not in new_device_unique_ids:
            device_registry.async_remove_device(old_entry.id)
            _LOGGER.debug("Remove device_registry: device_id=%s", old_entry.id)


async def async_unload_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
    """Unload the entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
