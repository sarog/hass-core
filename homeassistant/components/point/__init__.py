"""Support for Minut Point."""

import asyncio
from dataclasses import dataclass
from http import HTTPStatus
import logging

from aiohttp import ClientError, ClientResponseError, web
from pypoint import PointSession
import voluptuous as vol

from homeassistant.components import webhook
from homeassistant.components.application_credentials import (
    ClientCredential,
    async_import_client_credential,
)
from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
from homeassistant.const import (
    CONF_CLIENT_ID,
    CONF_CLIENT_SECRET,
    CONF_WEBHOOK_ID,
    Platform,
)
from homeassistant.core import DOMAIN as HOMEASSISTANT_DOMAIN, HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers import (
    aiohttp_client,
    config_entry_oauth2_flow,
    config_validation as cv,
    device_registry as dr,
)
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.dispatcher import (
    async_dispatcher_connect,
    async_dispatcher_send,
)
from homeassistant.helpers.entity import Entity
from homeassistant.helpers.event import async_track_time_interval
from homeassistant.helpers.issue_registry import IssueSeverity, async_create_issue
from homeassistant.helpers.typing import ConfigType
from homeassistant.util.dt import as_local, parse_datetime, utc_from_timestamp

from . import api
from .const import (
    CONF_WEBHOOK_URL,
    DOMAIN,
    EVENT_RECEIVED,
    POINT_DISCOVERY_NEW,
    SCAN_INTERVAL,
    SIGNAL_UPDATE_ENTITY,
    SIGNAL_WEBHOOK,
)

_LOGGER = logging.getLogger(__name__)

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]

type PointConfigEntry = ConfigEntry[PointData]

CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.Schema(
            {
                vol.Required(CONF_CLIENT_ID): cv.string,
                vol.Required(CONF_CLIENT_SECRET): cv.string,
            }
        )
    },
    extra=vol.ALLOW_EXTRA,
)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the Minut Point component."""
    if DOMAIN not in config:
        return True

    conf = config[DOMAIN]

    async_create_issue(
        hass,
        HOMEASSISTANT_DOMAIN,
        f"deprecated_yaml_{DOMAIN}",
        breaks_in_ha_version="2025.4.0",
        is_fixable=False,
        issue_domain=DOMAIN,
        severity=IssueSeverity.WARNING,
        translation_key="deprecated_yaml",
        translation_placeholders={
            "domain": DOMAIN,
            "integration_title": "Point",
        },
    )

    if not hass.config_entries.async_entries(DOMAIN):
        await async_import_client_credential(
            hass,
            DOMAIN,
            ClientCredential(
                conf[CONF_CLIENT_ID],
                conf[CONF_CLIENT_SECRET],
            ),
        )

        hass.async_create_task(
            hass.config_entries.flow.async_init(
                DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
            )
        )

    return True


async def async_setup_entry(hass: HomeAssistant, entry: PointConfigEntry) -> bool:
    """Set up Minut Point from a config entry."""

    if "auth_implementation" not in entry.data:
        raise ConfigEntryAuthFailed("Authentication failed. Please re-authenticate.")

    implementation = (
        await config_entry_oauth2_flow.async_get_config_entry_implementation(
            hass, entry
        )
    )
    session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
    auth = api.AsyncConfigEntryAuth(
        aiohttp_client.async_get_clientsession(hass), session
    )

    try:
        await auth.async_get_access_token()
    except ClientResponseError as err:
        if err.status in {HTTPStatus.UNAUTHORIZED, HTTPStatus.FORBIDDEN}:
            raise ConfigEntryAuthFailed from err
        raise ConfigEntryNotReady from err
    except ClientError as err:
        raise ConfigEntryNotReady from err

    point_session = PointSession(auth)

    client = MinutPointClient(hass, entry, point_session)
    hass.async_create_task(client.update())
    entry.runtime_data = PointData(client)

    await async_setup_webhook(hass, entry, point_session)
    # Entries are added in the client.update() function.

    return True


async def async_setup_webhook(
    hass: HomeAssistant, entry: PointConfigEntry, session: PointSession
) -> None:
    """Set up a webhook to handle binary sensor events."""
    if CONF_WEBHOOK_ID not in entry.data:
        webhook_id = webhook.async_generate_id()
        webhook_url = webhook.async_generate_url(hass, webhook_id)
        _LOGGER.debug("Registering new webhook at: %s", webhook_url)

        hass.config_entries.async_update_entry(
            entry,
            data={
                **entry.data,
                CONF_WEBHOOK_ID: webhook_id,
                CONF_WEBHOOK_URL: webhook_url,
            },
        )

    await session.update_webhook(
        webhook.async_generate_url(hass, entry.data[CONF_WEBHOOK_ID]),
        entry.data[CONF_WEBHOOK_ID],
        ["*"],
    )
    webhook.async_register(
        hass, DOMAIN, "Point", entry.data[CONF_WEBHOOK_ID], handle_webhook
    )


async def async_unload_entry(hass: HomeAssistant, entry: PointConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(
        entry, [*PLATFORMS, Platform.ALARM_CONTROL_PANEL]
    ):
        session: PointSession = entry.runtime_data.client
        if CONF_WEBHOOK_ID in entry.data:
            webhook.async_unregister(hass, entry.data[CONF_WEBHOOK_ID])
            await session.remove_webhook()
    return unload_ok


async def handle_webhook(
    hass: HomeAssistant, webhook_id: str, request: web.Request
) -> None:
    """Handle webhook callback."""
    try:
        data = await request.json()
        _LOGGER.debug("Webhook %s: %s", webhook_id, data)
    except ValueError:
        return

    if isinstance(data, dict):
        data["webhook_id"] = webhook_id
        async_dispatcher_send(hass, SIGNAL_WEBHOOK, data, data.get("hook_id"))
    hass.bus.async_fire(EVENT_RECEIVED, data)


class MinutPointClient:
    """Get the latest data and update the states."""

    def __init__(
        self, hass: HomeAssistant, config_entry: ConfigEntry, session: PointSession
    ) -> None:
        """Initialize the Minut data object."""
        self._known_devices: set[str] = set()
        self._known_homes: set[str] = set()
        self._hass = hass
        self._config_entry = config_entry
        self._is_available = True
        self._client = session

        async_track_time_interval(self._hass, self.update, SCAN_INTERVAL)

    async def update(self, *args):
        """Periodically poll the cloud for current state."""
        await self._sync()

    async def _sync(self):
        """Update local list of devices."""
        if not await self._client.update():
            self._is_available = False
            _LOGGER.warning("Device is unavailable")
            async_dispatcher_send(self._hass, SIGNAL_UPDATE_ENTITY)
            return

        async def new_device(device_id, platform):
            """Load new device."""
            async_dispatcher_send(
                self._hass, POINT_DISCOVERY_NEW.format(platform, DOMAIN), device_id
            )

        self._is_available = True
        for home_id in self._client.homes:
            if home_id not in self._known_homes:
                await self._hass.config_entries.async_forward_entry_setups(
                    self._config_entry, [Platform.ALARM_CONTROL_PANEL]
                )
                await new_device(home_id, "alarm_control_panel")
                self._known_homes.add(home_id)
        for device in self._client.devices:
            if device.device_id not in self._known_devices:
                await self._hass.config_entries.async_forward_entry_setups(
                    self._config_entry, PLATFORMS
                )
                for platform in PLATFORMS:
                    await new_device(device.device_id, platform)
                self._known_devices.add(device.device_id)
        async_dispatcher_send(self._hass, SIGNAL_UPDATE_ENTITY)

    def device(self, device_id):
        """Return device representation."""
        return self._client.device(device_id)

    def is_available(self, device_id):
        """Return device availability."""
        if not self._is_available:
            return False
        return device_id in self._client.device_ids

    async def remove_webhook(self):
        """Remove the session webhook."""
        return await self._client.remove_webhook()

    @property
    def homes(self):
        """Return known homes."""
        return self._client.homes

    async def async_alarm_disarm(self, home_id):
        """Send alarm disarm command."""
        return await self._client.alarm_disarm(home_id)

    async def async_alarm_arm(self, home_id):
        """Send alarm arm command."""
        return await self._client.alarm_arm(home_id)


class MinutPointEntity(Entity):  # pylint: disable=hass-enforce-class-module # see PR 118243
    """Base Entity used by the sensors."""

    _attr_should_poll = False

    def __init__(self, point_client, device_id, device_class) -> None:
        """Initialize the entity."""
        self._async_unsub_dispatcher_connect = None
        self._client = point_client
        self._id = device_id
        self._name = self.device.name
        self._attr_device_class = device_class
        self._updated = utc_from_timestamp(0)
        self._attr_unique_id = f"point.{device_id}-{device_class}"
        device = self.device.device
        self._attr_device_info = DeviceInfo(
            connections={(dr.CONNECTION_NETWORK_MAC, device["device_mac"])},
            identifiers={(DOMAIN, device["device_id"])},
            manufacturer="Minut",
            model=f"Point v{device['hardware_version']}",
            name=device["description"],
            sw_version=device["firmware"]["installed"],
            via_device=(DOMAIN, device["home"]),
        )
        if device_class:
            self._attr_name = f"{self._name} {device_class.capitalize()}"

    def __str__(self) -> str:
        """Return string representation of device."""
        return f"MinutPoint {self.name}"

    async def async_added_to_hass(self):
        """Call when entity is added to hass."""
        _LOGGER.debug("Created device %s", self)
        self._async_unsub_dispatcher_connect = async_dispatcher_connect(
            self.hass, SIGNAL_UPDATE_ENTITY, self._update_callback
        )
        await self._update_callback()

    async def async_will_remove_from_hass(self):
        """Disconnect dispatcher listener when removed."""
        if self._async_unsub_dispatcher_connect:
            self._async_unsub_dispatcher_connect()

    async def _update_callback(self):
        """Update the value of the sensor."""

    @property
    def available(self):
        """Return true if device is not offline."""
        return self._client.is_available(self.device_id)

    @property
    def device(self):
        """Return the representation of the device."""
        return self._client.device(self.device_id)

    @property
    def device_id(self):
        """Return the id of the device."""
        return self._id

    @property
    def extra_state_attributes(self):
        """Return status of device."""
        attrs = self.device.device_status
        attrs["last_heard_from"] = as_local(self.last_update).strftime(
            "%Y-%m-%d %H:%M:%S"
        )
        return attrs

    @property
    def is_updated(self):
        """Return true if sensor have been updated."""
        return self.last_update > self._updated

    @property
    def last_update(self):
        """Return the last_update time for the device."""
        return parse_datetime(self.device.last_update)


@dataclass
class PointData:
    """Point Data."""

    client: MinutPointClient
    entry_lock: asyncio.Lock = asyncio.Lock()
