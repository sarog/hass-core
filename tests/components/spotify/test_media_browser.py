"""Test the media browser interface."""

import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.spotify import DOMAIN
from homeassistant.components.spotify.browse_media import async_browse_media
from homeassistant.core import HomeAssistant
from homeassistant.setup import async_setup_component

from tests.common import MockConfigEntry


async def setup_integration(hass: HomeAssistant, config_entry: MockConfigEntry) -> None:
    """Fixture for setting up the component."""
    config_entry.add_to_hass(hass)

    await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done(wait_background_tasks=True)
    await async_setup_component(hass, DOMAIN, {})
    await hass.async_block_till_done(wait_background_tasks=True)


async def test_browse_media_root(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    spotify_setup,
) -> None:
    """Test browsing the root."""
    response = await async_browse_media(hass, None, None)
    assert response.as_dict() == snapshot


async def test_browse_media_categories(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    spotify_setup,
) -> None:
    """Test browsing categories."""
    response = await async_browse_media(
        hass, "spotify://library", "spotify://01J5TX5A0FF6G5V0QJX6HBC94T"
    )
    assert response.as_dict() == snapshot


@pytest.mark.parametrize(
    ("config_entry_id"), [("01J5TX5A0FF6G5V0QJX6HBC94T"), ("32oesphrnacjcf7vw5bf6odx3")]
)
async def test_browse_media_playlists(
    hass: HomeAssistant,
    snapshot: SnapshotAssertion,
    config_entry_id: str,
    spotify_setup,
) -> None:
    """Test browsing playlists for the two config entries."""
    response = await async_browse_media(
        hass,
        "spotify://current_user_playlists",
        f"spotify://{config_entry_id}/current_user_playlists",
    )
    assert response.as_dict() == snapshot
