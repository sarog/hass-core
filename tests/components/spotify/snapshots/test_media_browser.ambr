# serializer version: 1
# name: test_browse_media_categories
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_playlists',
        'media_content_type': 'spotify://current_user_playlists',
        'thumbnail': None,
        'title': 'Playlists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ARTIST: 'artist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_followed_artists',
        'media_content_type': 'spotify://current_user_followed_artists',
        'thumbnail': None,
        'title': 'Artists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ALBUM: 'album'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_albums',
        'media_content_type': 'spotify://current_user_saved_albums',
        'thumbnail': None,
        'title': 'Albums',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_tracks',
        'media_content_type': 'spotify://current_user_saved_tracks',
        'thumbnail': None,
        'title': 'Tracks',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PODCAST: 'podcast'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_saved_shows',
        'media_content_type': 'spotify://current_user_saved_shows',
        'thumbnail': None,
        'title': 'Podcasts',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_recently_played',
        'media_content_type': 'spotify://current_user_recently_played',
        'thumbnail': None,
        'title': 'Recently played',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ARTIST: 'artist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_top_artists',
        'media_content_type': 'spotify://current_user_top_artists',
        'thumbnail': None,
        'title': 'Top Artists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_top_tracks',
        'media_content_type': 'spotify://current_user_top_tracks',
        'thumbnail': None,
        'title': 'Top Tracks',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.GENRE: 'genre'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/categories',
        'media_content_type': 'spotify://categories',
        'thumbnail': None,
        'title': 'Categories',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/featured_playlists',
        'media_content_type': 'spotify://featured_playlists',
        'thumbnail': None,
        'title': 'Featured Playlists',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': <MediaClass.ALBUM: 'album'>,
        'media_class': <MediaClass.DIRECTORY: 'directory'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/new_releases',
        'media_content_type': 'spotify://new_releases',
        'thumbnail': None,
        'title': 'New Releases',
      }),
    ]),
    'children_media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/library',
    'media_content_type': 'spotify://library',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Media Library',
  })
# ---
# name: test_browse_media_playlists
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/spotify:playlist:unique_identifier_00',
        'media_content_type': 'spotify://playlist',
        'thumbnail': None,
        'title': 'Playlist1',
      }),
    ]),
    'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_playlists',
    'media_content_type': 'spotify://current_user_playlists',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Playlists',
  })
# ---
# name: test_browse_media_playlists[01J5TX5A0FF6G5V0QJX6HBC94T]
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/spotify:playlist:unique_identifier_00',
        'media_content_type': 'spotify://playlist',
        'thumbnail': None,
        'title': 'Playlist1',
      }),
    ]),
    'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://01j5tx5a0ff6g5v0qjx6hbc94t/current_user_playlists',
    'media_content_type': 'spotify://current_user_playlists',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Playlists',
  })
# ---
# name: test_browse_media_playlists[32oesphrnacjcf7vw5bf6odx3]
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': True,
        'children_media_class': <MediaClass.TRACK: 'track'>,
        'media_class': <MediaClass.PLAYLIST: 'playlist'>,
        'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3/spotify:playlist:unique_identifier_00',
        'media_content_type': 'spotify://playlist',
        'thumbnail': None,
        'title': 'Playlist1',
      }),
    ]),
    'children_media_class': <MediaClass.PLAYLIST: 'playlist'>,
    'media_class': <MediaClass.DIRECTORY: 'directory'>,
    'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3/current_user_playlists',
    'media_content_type': 'spotify://current_user_playlists',
    'not_shown': 0,
    'thumbnail': None,
    'title': 'Playlists',
  })
# ---
# name: test_browse_media_root
  dict({
    'can_expand': True,
    'can_play': False,
    'children': list([
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': None,
        'media_class': <MediaClass.APP: 'app'>,
        'media_content_id': 'spotify://01J5TX5A0FF6G5V0QJX6HBC94T',
        'media_content_type': 'spotify://library',
        'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
        'title': 'spotify_1',
      }),
      dict({
        'can_expand': True,
        'can_play': False,
        'children_media_class': None,
        'media_class': <MediaClass.APP: 'app'>,
        'media_content_id': 'spotify://32oesphrnacjcf7vw5bf6odx3',
        'media_content_type': 'spotify://library',
        'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
        'title': 'spotify_2',
      }),
    ]),
    'children_media_class': <MediaClass.APP: 'app'>,
    'media_class': <MediaClass.APP: 'app'>,
    'media_content_id': 'spotify://',
    'media_content_type': 'spotify',
    'not_shown': 0,
    'thumbnail': 'https://brands.home-assistant.io/_/spotify/logo.png',
    'title': 'Spotify',
  })
# ---
